The DSA `sparse_mla_fwd_interface` operator fails accuracy tests

### Describe the bug

The `sparse_mla_fwd_interface` operator always fail the accuracy tests with following errors:

```
platform linux -- Python 3.11.9, pytest-8.2.1, pluggy-1.5.0
rootdir: /root/actions-runner/_work/FlagGems/FlagGems
configfile: pytest.ini
plugins: md-report-0.7.0
collected 166 items
tests/test_DSA/test_sparse_mla_ops.py FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF.
=================================== FAILURES ===================================
__________ test_sparse_mla_forward[dtype0-64-512-576-1-64-1024-64-1] ___________
batch_size = 1, seq_len_q = 64, seq_len_kv = 1024, num_heads = 64
num_kv_heads = 1, qk_dim = 576, d_v = 512, topk = 64, dtype = torch.bfloat16
    @pytest.mark.sparse_mla_fwd_interface
    @pytest.mark.parametrize("batch_size", [1])
    @pytest.mark.parametrize("seq_len_q", [64, 128, 512])
    @pytest.mark.parametrize("seq_len_kv", [1024, 2048, 4096])
    @pytest.mark.parametrize("num_heads", [64, 128, 256])
    @pytest.mark.parametrize("num_kv_heads", [1, 2])
    @pytest.mark.parametrize("qk_dim", [576])  # Your operator is fixed at 576
    @pytest.mark.parametrize("d_v", [512])  # Output dimension
    @pytest.mark.parametrize("topk", [64, 128, 256])
    @pytest.mark.parametrize("dtype", [torch.bfloat16])
    def test_sparse_mla_forward(
        batch_size: int,
        seq_len_q: int,
        seq_len_kv: int,
        num_heads: int,
        num_kv_heads: int,
        qk_dim: int,
        d_v: int,
        topk: int,
        dtype: torch.dtype,
    ):
        """Sparse MLA forward propagation test"""
        # Skip unsupported cases
        if num_heads % num_kv_heads != 0:
            pytest.skip("num_heads must be divisible by num_kv_heads")
    
        if topk > seq_len_kv:
            pytest.skip("topk cannot be larger than seq_len_kv")
    
        # Create input
        q, kv, indices = make_sparse_mla_input(
            batch_size,
            seq_len_q,
            seq_len_kv,
            num_heads,
            num_kv_heads,
            qk_dim,
            topk,
            dtype,
            device,
        )
    
        # Reference implementation
        ref_q = to_reference(q, False)
        ref_kv = to_reference(kv, False)
        ref_indices = to_reference(indices, False)
    
>       ref_output = reference_sparse_mla_implementation(
            ref_q, ref_kv, ref_indices, d_v=d_v
        )
tests/test_DSA/test_sparse_mla_ops.py:155: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_DSA/test_sparse_mla_ops.py:105: in reference_sparse_mla_implementation
    return ref_sparse_mla_fwd_interface(q, kv, indices, sm_scale=sm_scale, d_v=d_v)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
q = tensor([[[[ 1.9434e-01,  2.1562e+00, -1.7188e-01,  ...,  2.9492e-01,
           -1.0469e+00, -8.8672e-01],
          [...703e-01],
          [-5.3516e-01, -4.4141e-01,  2.3535e-01,  ...,  5.6396e-02,
           -1.2109e+00, -2.0156e+00]]]])
kv = tensor([[[[-0.6992, -1.8672, -0.8828,  ..., -0.5039,  0.8438, -0.9102]],
         [[-1.0312, -0.9453,  2.1094,  ..., ... 0.6641,  ..., -1.3125,  0.4043,  0.0212]],
         [[-1.2734,  0.7188,  0.4121,  ...,  0.6914, -0.7773,  1.4141]]]])
indices = tensor([[[[   0, 1024, 1024,  ..., 1024, 1024, 1024],
          [   0, 1024, 1024,  ..., 1024, 1024, 1024],
          ...[   7,   41,   13,  ...,   28, 1024, 1024],
          [  56,   39,   57,  ...,   19,   16, 1024]]]], dtype=torch.int32)
sm_scale = None, is_casual = True, d_v = 512
    def ref_sparse_mla_fwd_interface(
        q, kv, indices, sm_scale=None, is_casual=True, d_v=512
    ):
        q = q.float()
        kv = kv.float()
        indices = indices.transpose(1, 2)
        b, sq, h, dim_q = q.shape
        b, sk, g, _ = kv.shape
    
        assert kv.shape[-1] == 576, "you should assign dim otherwise"
        dim = 512
        k = kv
        v = kv[..., :dim]
    
        b, _, _, dim_v = v.shape
        g_index = g
        h_index = h // g
        compressed_casual_mask = torch.arange(0, sq, dtype=torch.int32, device="cuda").view(
            -1, 1
        ) >= torch.arange(1 - 1, sk * 1, 1, dtype=torch.int32, device="cuda").view(1, -1)
    
        mask = q.new_zeros(b, g_index, sq, sk + 1, dtype=torch.bool).scatter(
            3, indices.long(), 1
        )
        mask = mask[..., :-1]
>       mask = mask & compressed_casual_mask.view(1, 1, sq, sk)
E       RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
tests/test_DSA/torch_src/sparse_mla_fwd.py:30: RuntimeError

```

--- a/tests/test_DSA/test_sparse_mla_ops.py
+++ b/tests/test_DSA/test_sparse_mla_ops.py
@@ -106,7 +106,6 @@ def reference_sparse_mla_implementation(q, kv, indices, sm_scale=None, d_v=512):
 
 
 @pytest.mark.sparse_mla_fwd_interface
-@pytest.mark.skip("#2872: RuntimeError")
 @pytest.mark.parametrize("batch_size", [1])
 @pytest.mark.parametrize("seq_len_q", [64, 128, 512])
 @pytest.mark.parametrize("seq_len_kv", [1024, 2048, 4096])
@@ -165,7 +164,6 @@ def test_sparse_mla_forward(
 
 
 @pytest.mark.sparse_mla_fwd_interface
-@pytest.mark.skip("#2872: RuntimeError")
 @pytest.mark.parametrize(
     "config",
     [
@@ -226,7 +224,6 @@ def test_sparse_mla_forward_edge_cases(config):
 
 
 # Device compatibility test
-@pytest.mark.skip("#2872: RuntimeError")
 @pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA device required")
 @pytest.mark.sparse_mla_fwd_interface
 def test_sparse_mla_device_compatibility():

--- a/tests/test_DSA/torch_src/sparse_mla_fwd.py
+++ b/tests/test_DSA/torch_src/sparse_mla_fwd.py
@@ -19,9 +19,12 @@ def ref_sparse_mla_fwd_interface(
     b, _, _, dim_v = v.shape
     g_index = g
     h_index = h // g
-    compressed_casual_mask = (
-        torch.arange(0, sq, dtype=torch.int32, device=q.device).view(-1, 1)
-        >= torch.arange(1 - 1, sk * 1, 1, dtype=torch.int32, device=q.device).view(1, -1)
+    compressed_casual_mask = torch.arange(
+        0, sq, dtype=torch.int32, device=q.device
+    ).view(-1, 1) >= torch.arange(
+        1 - 1, sk * 1, 1, dtype=torch.int32, device=q.device
+    ).view(
+        1, -1
     )
 
     mask = q.new_zeros(b, g_index, sq, sk + 1, dtype=torch.bool).scatter(

--- a/tests/test_DSA/torch_src/sparse_mla_fwd.py
+++ b/tests/test_DSA/torch_src/sparse_mla_fwd.py
@@ -19,9 +19,9 @@ def ref_sparse_mla_fwd_interface(
     b, _, _, dim_v = v.shape
     g_index = g
     h_index = h // g
-    compressed_casual_mask = torch.arange(0, sq, dtype=torch.int32, device="cuda").view(
+    compressed_casual_mask = torch.arange(0, sq, dtype=torch.int32, device=q.device).view(
         -1, 1
-    ) >= torch.arange(1 - 1, sk * 1, 1, dtype=torch.int32, device="cuda").view(1, -1)
+    ) >= torch.arange(1 - 1, sk * 1, 1, dtype=torch.int32, device=q.device).view(1, -1)
 
     mask = q.new_zeros(b, g_index, sq, sk + 1, dtype=torch.bool).scatter(
         3, indices.long(), 1

--- a/tests/test_DSA/torch_src/sparse_mla_fwd.py
+++ b/tests/test_DSA/torch_src/sparse_mla_fwd.py
@@ -19,9 +19,10 @@ def ref_sparse_mla_fwd_interface(
     b, _, _, dim_v = v.shape
     g_index = g
     h_index = h // g
-    compressed_casual_mask = torch.arange(0, sq, dtype=torch.int32, device=q.device).view(
-        -1, 1
-    ) >= torch.arange(1 - 1, sk * 1, 1, dtype=torch.int32, device=q.device).view(1, -1)
+    compressed_casual_mask = (
+        torch.arange(0, sq, dtype=torch.int32, device=q.device).view(-1, 1)
+        >= torch.arange(1 - 1, sk * 1, 1, dtype=torch.int32, device=q.device).view(1, -1)
+    )
 
     mask = q.new_zeros(b, g_index, sq, sk + 1, dtype=torch.bool).scatter(
         3, indices.long(), 1

--- a/tools/test-op.sh
+++ b/tools/test-op.sh
@@ -40,12 +40,15 @@ TEST_CASES=()
 PERF_TEST_CASES=()
 TEST_CASES_CPU=()
 for item in $CHANGED_FILES; do
+  file_name=$(basename "$item")
   case $item in
     tests/test_quant.py)
       # skip because it always fail
       ;;
-    tests/test*)
-      TEST_CASES+=($item)
+    tests/*.py)
+      if [[ "$file_name" == test*.py ]]; then
+        TEST_CASES+=($item)
+      fi
       ;;
     benchmark/test*)
       PERF_TEST_CASES+=($item)
@@ -62,7 +65,11 @@ for item in $CHANGED_FILES; do
   done
   if (( $found == 0 )); then
     case $item in
-      tests/*) TEST_CASES_CPU+=($item) ;;
+      tests/*.py)
+        if [[ "$file_name" == test*.py ]]; then
+          TEST_CASES_CPU+=($item)
+        fi
+        ;;
     esac
   fi
 done

